BYOC Blog Post (#12)
apache/incubator-tvm-site: _posts/2020-07-13-how-to-bring-your-own-codegen-to-tvm.md

If your compilation flow falls into this case, then we recommend reading all the rest sections in this post but skipping [Bring DNNL to TVM: C Source Codegen](#bring-dnnl-to-tvm-c-source-codegen).

**Generate assembly code and compile it to an executable binary**:
If you do not have an end-to-end execution framework for your platform like the previous case, you may have a compiler to compile a program in assembly code of your ISA. In order to feed the assembly code to your compiler, you will need a codegen to generate and optimize the assembly code from a Relay graph.

If your compilation flow falls into this case, then we recommend reading all the rest sections in this post but skipping [Bring DNNL to TVM: JSON Codegen/Runtime](#bring-dnnl-to-tvm-json-codegenruntime).

## How BYOC Works

> We then briefly explain how BYOC framework work. For more detail explanations of underlying framework components and their implementations, please refer to the [developer document]([https://tvm.apache.org/docs/dev/relay_bring_your_own_codegen.html](https://tvm.apache.org/docs/dev/relay_bring_your_own_codegen.html)). In short, given a Relay graph in Figure 1, BYOC framework does the following steps:

Reviewer: ```suggestion
We then briefly explain how BYOC framework works. For more detail explanations of underlying framework components and their implementations, please refer to the [developer document]([https://tvm.apache.org/docs/dev/relay_bring_your_own_codegen.html](https://tvm.apache.org/docs/dev/relay_bring_your_own_codegen.html)). In short, given a Relay graph in Figure 1, BYOC framework does the following steps:
```

{:center: style="text-align: center"}
![The original Relay graph](/images/bring-your-own-codegen/original_graph.png){: width="50%"}
{:center}
<center>
Figure 1: The Original Relay Graph.
</center> <p></p>

### 1. Graph Annotation
Taking a user-provided Relay graph, our first step is to annotate the nodes that potentially can be offloaded to your accelerator in the graph. You will need to follow [Bring DNNL to TVM: Annotation Rules](#bring-dnnl-to-tvm-annotation-rules) to implement a whitelist of supported operators, or a graph pattern list of customized composite operators. An example annotation result is shown in Figure 2.

{:center: style="text-align: center"}
![The Graph with Annotations](/images/bring-your-own-codegen/after_annotation.png){: width="50%"}
{:center}
<center>
Figure 2: The Graph with Annotations.
</center> <p></p>

### 2. Graph Transformation
The second step is to transform and optimize the graph based on the annotations. Specifically, BYOC performs the following transformations.

> **2.1: Merge compiler region**: As can be seen in Figure 2, we now have many "regions" in the graph that can be offloaded to your accelerator, but some of them can actually be merged to reduce the data transfer and kernel launching overhead. Accordingly, step 2.1 performs greedy algorithm to merge those regions as many as possible while guaranteeing the functional correctness. The result is depicted in Figure 3.

Reviewer: ```suggestion
**2.1: Merge compiler region**: As can be seen in Figure 2, we now have many "regions" in the graph that can be offloaded to your accelerator, but some of them can actually be merged to reduce the data transfer and kernel launching overhead. Accordingly, step 2.1 uses a greedy algorithm to merge as many of those regions as possible while guaranteeing the functional correctness. The result is depicted in Figure 3.
```

{:center: style="text-align: center"}
![After Merging Compiler Regions](/images/bring-your-own-codegen/after_merging_regions.png){: width="50%"}
{:center}
<center>
Figure 3: After Merging Compiler Regions.
</center> <p></p>

**2.2: Partition Graph**: For each region from the previous step, we create a Relay function with an attribute `Compiler` to indicate that this Relay function should be entirely offloaded to your accelerator, as shown in Figure 4.

{:center: style="text-align: center"}
![After Graph Partitioning](/images/bring-your-own-codegen/after_partitioning.png){: width="50%"}
{:center}
<center>
Figure 4: After Graph Partitioning.
</center> <p></p>

### 3. Code Generation
Now we know which part of the Relay graph should be offloaded. In this step, we sequentially send every Relay function with `Compiler=your_accelerator` to your codegen. Your codegen should compile the Relay function to the form that matches your own compilation flow. It can be either C source code or any text formats.

> Finally, all compiled functions will be serialized along with other non-offloaded Relay functions to a single `.so` file by the TVM `export_library` Python API. In other words, user will get only one `.so` file after running this flow.

Reviewer: ```suggestion
Finally, all compiled functions will be serialized along with other non-offloaded Relay functions to a single `.so` file by the TVM `export_library` Python API. In other words, the user will get only one `.so` file after running this flow.
```

### 4. Runtime
You may also need to implement a runtime to initialize your graph engine (if applicable) and execute the compiled functions. During the inference, TVM runtime (i.e., graph runtime or VM) will leverage your runtime to invoke the offloaded functions when the TVM runtime encounters the corresponding function call in Figure 4. Your runtime is responsible for launching the compiled function with the given input tensor arrays and filling in the results to the output tensor arrays.

> In the rest of this post, we use DNNL as an example to demonstrate how to achieve the above workflow using BYOC framework. Please note that all referred code and line number in this post are based on the TVM repository's master branch commit [8a0249c](https://github.com/apache/incubator-tvm/tree/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8).

Reviewer: ```suggestion
In the rest of this post, we use DNNL as an example to demonstrate how to achieve the above workflow using the BYOC framework. Please note that all referred code and line number in this post are based on the TVM repository's master branch commit [8a0249c](https://github.com/apache/incubator-tvm/tree/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8).
```

## Bring DNNL to TVM: Annotation Rules

> The BYOC framework provides two approaches for you to describe the supported operators and patterns. You can use both of them simultaneously. In this section, we use DNNL as an example to show how to make use of them. The complete implementation is available [here](https://github.com/apache/incubator-tvm/blob/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8/python/tvm/relay/op/contrib/dnnl.py). Note that we put the annotation rules for your codegens under `python/tvm/relay/op/contrib/your_codegen_name.py`.

Reviewer: ```suggestion
The BYOC framework provides two approaches for you to describe the supported operators and patterns. You can use both of them simultaneously. In this section, we use DNNL as an example to show how to make use of them. The complete implementation is available [here](https://github.com/apache/incubator-tvm/blob/8a0249cd4d12a2eb1a4e7a692a9265bc63fec5c8/python/tvm/relay/op/contrib/dnnl.py). Note that we put the annotation rules for your codegen under `python/tvm/relay/op/contrib/your_codegen_name.py`.
```
